File: aat-lib/src/main/java/ch/bailu/aat_lib/map/tile/TileObjectCache.java
package ch.bailu.aat_lib.map.tile;

import org.mapsforge.core.model.Tile;

import java.io.Closeable;

import ch.bailu.aat_lib.service.cache.LockCache;
import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.service.cache.ObjTile;

public class TileObjectCache implements Closeable {


    private final static int INITIAL_CAPACITY = 5;

    private final LockCache<ObjTile> tiles = new LockCache<>(INITIAL_CAPACITY);

    public final static TileObjectCache NULL = new TileObjectCache() {
        @Override
        public void put(ObjTile t) {

        }
    };


    /**
     * Non synchronized function. This works because LockCache can only grow.
     *
     */
    public boolean isInCache(String string) {
        if (string != null) {
            int hash = string.hashCode();

            for (int i = 0; i < tiles.size(); i++) {
                ObjTile o = tiles.get(i);

                if (o != null && hash == tiles.get(i).hashCode()) {
                    return true;
                }
            }

        }
        return false;
    }


    public synchronized ObjTile get(String string) {
        for (int i = 0; i<tiles.size(); i++) {
            if (tiles.get(i).toString().equals(string)) {
                return tiles.use(i);
            }
        }
        return null;
    }


    public synchronized ObjTile get(Tile tile) {
        for (int i = 0; i<tiles.size(); i++) {
            if (compare(tile, tiles.get(i).getTile())) {
                return tiles.use(i);
            }
        }
        return null;
    }

    public static boolean compare(Tile a, Tile b) {
        return a.tileX == b.tileX && a.tileY == b.tileY && a.zoomLevel == b.zoomLevel;
    }


    public synchronized void put(ObjTile handle) {
        tiles.add(handle);
    }

    @Override
    public synchronized void close() {
        tiles.close();
    }


    public synchronized void reDownloadTiles(AppContext sc) {
        for (int i = 0; i<tiles.size(); i++) {
            tiles.get(i).reDownload(sc);
        }
    }


    public synchronized void reset() {
        tiles.reset();
    }


    public synchronized void setCapacity(int capacity) {
        tiles.ensureCapacity(capacity);
    }


    public synchronized int size() {
        return tiles.size();
    }


    public synchronized boolean isReadyAndLoaded() {
        for (int i = 0; i<tiles.size(); i++) {
            //AppLog.d(this, tiles.get(i).getID());
            if (tiles.get(i) != null) {
                if (!tiles.get(i).isReadyAndLoaded())
                    return false;
            }
        }
        return true;
    }

    public synchronized boolean isLoaded() {
        for (int i = 0; i<tiles.size(); i++) {
            if (tiles.get(i) != null) {
                if (!tiles.get(i).isLoaded())
                    return false;
            }
        }
        return true;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/tile/TileObjectCache.kt
package ch.bailu.aat_lib.map.tile

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.service.cache.LockCache
import ch.bailu.aat_lib.service.cache.ObjTile
import org.mapsforge.core.model.Tile
import java.io.Closeable

open class TileObjectCache : Closeable {
    private val tiles = LockCache<ObjTile>(INITIAL_CAPACITY)

    /**
     * Non synchronized function. This works because LockCache can only grow.
     * @param fileId id (path) of file that might be inside the cache
     * @return true if object corresponding to  fileId is in cache else return false
     */
    fun isInCache(fileId: String): Boolean {
        for (i in 0 until tiles.size()) {
            val tile = tiles[i]
            if (tile is ObjTile && fileId == tile.id) {
                return true
            }
        }
        return false
    }

    /**
     * Get the tile with specific zoom level and coordinates from the cache
     * If it can find the tile inside the cache it sets the access time to now.
     *
     * @param tile Coordinates and zoom level of the requested tile
     * @return A tile from the cache or null if it was not in the cache
     */
    @Synchronized
    operator fun get(tile: Tile): ObjTile? {
        for (i in 0 until tiles.size()) {
            val other = tiles[i]
            if (other is ObjTile && compare(other.getTile(), tile)) {
                return tiles.use(i)
            }
        }
        return null
    }

    @Synchronized
    open fun put(handle: ObjTile) {
        tiles.add(handle)
    }

    @Synchronized
    override fun close() {
        tiles.close()
    }

    @Synchronized
    fun reDownloadTiles(sc: AppContext) {
        for (i in 0 until tiles.size()) {
            tiles[i]?.reDownload(sc)
        }
    }

    @Synchronized
    fun reset() {
        tiles.reset()
    }

    @Synchronized
    fun setCapacity(capacity: Int) {
        tiles.ensureCapacity(capacity)
    }

    @Synchronized
    fun size(): Int {
        return tiles.size()
    }

    @get:Synchronized
    val isReadyAndLoaded: Boolean
        get() {
            for (i in 0 until tiles.size()) {
                if (tiles[i] != null) {
                    if (tiles[i]?.isReadyAndLoaded != true) return false
                }
            }
            return true
        }

    companion object {
        private const val INITIAL_CAPACITY = 5
        val NULL: TileObjectCache = object : TileObjectCache() {
            override fun put(handle: ObjTile) {}
        }

        fun compare(a: Tile, b: Tile): Boolean {
            return a.tileX == b.tileX && a.tileY == b.tileY && a.zoomLevel == b.zoomLevel
        }
    }
}
